File: Week 2/introduction_starter_files/fibonacci_huge/fibonacci_huge.py
# Uses python3
import sys

def get_fibonacci_huge_naive(n, m):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % m
	
	
def get_fibonacci_huge_fast(n,m):
	if (n<=1):
		return n;
	
	sequence =[];
	sequence.append(0);
	sequence.append(1);
	prev =0;
	curr = 1;
	
	i=2
	while i <n+1:
		fib=(curr+prev)%m;
		prev = curr;
		curr = fib;
		sequence.append(fib));
		if((fib%m) ==0):
			fib=(curr+prev;)%m
			#print("i ",i)
			
			if(fib%m == 1):
				sequence.pop();
				break;
	
		i+=1;
		
	#print(sequence);
	#print(len(sequence))
	val = n % len(sequence);#print(val)
	result = sequence[val];
	return result;

if __name__ == '__main__':
	input = sys.stdin.read();
	n, m = map(int, input.split())
	#print(get_fibonacci_huge_naive(n, m))
	print(get_fibonacci_huge_fast(n, m))

File: Week 2/introduction_starter_files/fibonacci_huge/fibonacci_huge.cpp
#include <iostream>

long long get_fibonacci_huge_naive(long long n, long long m) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    return current % m;
}

long long get_fibonacci_huge_fast(long long n,long long m){
	if (n<=1)
		return n;
	
	long long sequence[6*m];
	sequence[0]=0;
	sequence[1]=1;
	long long prev =0;
	long long curr = 1;
	long long val;
    long long result;
	long long i=2;
    long long fib;
	for(;i<n+1;i++){
		fib=(curr+prev)%m;
		prev = curr;
		curr = fib;
		sequence[i] = fib;
        /*if(sequence[i] <0)
        {
            std::cout<<" i at error " << i<<'\n';
            std::cout<<" fib at error " << fib<<'\n';
            std::cout<<" prev at error " << prev<<'\n';
        }*/
		if((fib) ==0){
			fib=(curr+prev)%m;
			
			if(fib == 1){
				i--;
				break;
			}
		}
	}
	//print(sequence);
    //for(long long j=0;j<=i;j++)
    //    std::cout<<sequence[j]<<" ";
    
	//print(len(sequence))
    i = i+1;
    //std::cout<<"i "<<i<<'\n';
	val = n %i ;
    //std::cout<<"val "<<val<<'\n';
	result = sequence[val];
	return result;
	
}



int main() {
    long long n, m;
    std::cin >> n >> m;
    //std::cout << get_fibonacci_huge_naive(n, m) << '\n';
	std::cout << get_fibonacci_huge_fast(n, m) << '\n';
}
